docs/build.py

```diff
@@ -6,249 +6,390 @@
+    examples_path = os.path.join(project_path, 'examples')
+    lvgl_src_path = os.path.join(project_path, 'src')
+    latex_output_path = os.path.join(temp_directory, 'out_latex')
+    pdf_src_file = os.path.join(latex_output_path, 'LVGL.pdf')
+    pdf_dst_file = os.path.join(temp_directory, 'LVGL.pdf')
+    html_src_path = temp_directory
+    html_dst_path = args.output_path
 
+    lvgl_header_path = os.path.join(project_path, 'lvgl.h')
+    lv_config_path = os.path.join(base_path, 'lv_conf.h')
 
-def cmd(s):
-    print("")
-    print(s)
-    print("-------------------------------------")
+    os.chdir(base_path)
 
-    result = os.system(s)
-    if result != 0:
-        print("Exit build due to previous error")
-        sys.exit(result)
+    if temp_path != temp_directory:
+        if not os.path.exists(temp_directory):
+            os.makedirs(temp_directory)
+            temp_path_exists = False
+        else:
+            temp_path_exists = True
 
+        shutil.rmtree(temp_path)
+    else:
+        temp_path_exists = False
 
-# Get the current branch name
-status, br = subprocess.getstatusoutput("git branch --show-current")
-_, gitcommit = subprocess.getstatusoutput("git rev-parse HEAD")
-br = re.sub('\* ', '', br)
+    def cmd(s):
+        print('running command: ', s)
 
+        result = os.system(s)
+        if result != 0:
+            print("Exit build due to previous error")
+            sys.exit(result)
 
-urlpath = re.sub('release/', '', br)
+    # Get the current branch name
+    status, br = subprocess.getstatusoutput("git branch --show-current")
+    _, gitcommit = subprocess.getstatusoutput("git rev-parse HEAD")
+    br = re.sub('\* ', '', br)  # NOQA
 
-os.environ['LVGL_URLPATH'] = urlpath
-os.environ['LVGL_GITCOMMIT'] = gitcommit
+    urlpath = re.sub('release/', '', br)
 
+    os.environ['LVGL_URLPATH'] = urlpath
+    os.environ['LVGL_GITCOMMIT'] = gitcommit
 
-lang = "en"
-print("")
-print("****************")
-print("Building")
-print("****************")
+    if clean:
+        print('cleaning...')
+        if os.path.exists(html_dst_path):
+            shutil.rmtree(html_dst_path)
 
-if clean:
-    print('cleaning...')
-    # api_path = os.path.join(dname, 'API')
-    # xml_path = os.path.join(dname, 'xml')
-    # doxy_path = os.path.join(dname, 'doxygen_html')
+        if temp_path_exists:
+            try:
+                shutil.rmtree(temp_directory)
+                os.mkdir(temp_directory)
+            except OSError:
+                raise RuntimeError(
+                    f'Unable to clean build path "{temp_directory}"'
+                )
+
+    print('generating lv_conf.h file...')
+    config_builder.run()
+    print('copying documentation files to build directory...')
+    shutil.copytree('.', temp_directory, dirs_exist_ok=True)
+    print('copying examkples to build directory...')
+    shutil.copytree(examples_path, os.path.join(temp_directory, 'examples'))
+
+    def iter_rst_files(p):
```

Comment:
I'll pin my comment to this line, but it's a more generic feedback.

I feel like this script is getting a bit messy and hard to understand for future contributors. E.g.:
- function definitions and instructions mixing in the execution flow
- high level nesting and code repetition
- the idea of a script reading its own stdout / stderr feels a awkward

My suggestions would be:
- use `os.walk()`
- move class and method definitions to the beginning of file / separate module
- move "clean", "doxy", "pdf" and "html" phases to standalone methods which can be called or not depending on the cli arguments
- I'd expect a flag to exist which makes doxygen/sphinx to return an error code upon any warning. We should pass that, and watch the return value instead of monitoring the output. If that's not possible for some reason, we could still use an external script - bash might be a better fit for this role -  to catch keywords in std stream.

I'm happy to make these changes in an upcoming PR if you want to move forward with this one

Reply:
If you believe you are able to write a better script then have at it. Now remember this script is not just used to build the documentation so make sure that everything else still functions properly with the script you write.